### PR TITLE
WIP: Rate limit discussions POST routes

### PR DESCRIPTION
**Pull request by abulte**
Rebased on #2182.

This lets us rate limit those API endpoints:
- post a discussion
- post a comment in a discussion

The rate limit is disabled by default and its counter is stored in Redis (can be in memory or memcached). The rate is configurable (defaults to 5 per minute for comments).

The rate limiting logic is easily reusable on other endpoints if needed.

## TODO

- [x] tests
- [x] assess performance on next.data.gouv.fr before/after

### CODE REVIEW

**Comment by abulte** [comment]
poke @odtvince

**Comment by abulte** [comment]
Results a little load testing on next.data.gouv.fr, using locust.io with the following script https://gist.github.com/abulte/79311dc7eef10dcfc79a5bf12e0fa0a3 (create 1 dataset, foreach create 10 discussions, foreach create 10 comments, loop for 1 minute) and the following command `locust --no-web -c 1 -r 1 -t 1m`.

## Rate Limit off

```
Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 POST /api/1/datasets/                                               5    130    150    150    390    390    390    390    390    390
 POST /api/1/discussions/                                           41    110    120    130    140    180    220    390    390    390
 POST /api/1/discussions/5cf7d13cdc47091405a78001/                  10    130    130    130    130    230    230    230    230    230
 POST /api/1/discussions/5cf7d13ddc470914041794e7/                  10    100    100    110    130    520    520    520    520    520
 POST /api/1/discussions/5cf7d13fdc47095ca2366740/                  10    130    150    150    150    200    200    200    200    200
 POST /api/1/discussions/5cf7d141dc47091406fe0662/                  10    110    110    120    120    610    610    610    610    610
 POST /api/1/discussions/5cf7d142dc47097e573fa12e/                  10    100    110    110    110    250    250    250    250    250
 POST /api/1/discussions/5cf7d144dc47091405a78002/                  10    100    110    110    120    130    130    130    130    130
 POST /api/1/discussions/5cf7d145dc470914041794e8/                  10     98     99    110    110    110    110    110    110    110
 POST /api/1/discussions/5cf7d146dc47095ca2366741/                  10    100    110    110    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d147dc47091406fe0663/                  10    100    100    110    110    890    890    890    890    890
 POST /api/1/discussions/5cf7d149dc47097e573fa12f/                  10    110    110    120    130    140    140    140    140    140
 POST /api/1/discussions/5cf7d14bdc470914041794e9/                  10    100    100    110    130    130    130    130    130    130
 POST /api/1/discussions/5cf7d14cdc47095ca2366742/                  10     97     98     99    110    150    150    150    150    150
 POST /api/1/discussions/5cf7d14ddc47091406fe0664/                  10    100    100    110    110    270    270    270    270    270
 POST /api/1/discussions/5cf7d14edc47097e573fa130/                  10    100    100    100    110    110    110    110    110    110
 POST /api/1/discussions/5cf7d14fdc47091405a78004/                  10     98     98    100    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d150dc470914041794ea/                  10    110    110    110    130    130    130    130    130    130
 POST /api/1/discussions/5cf7d152dc47095ca2366743/                  10    130    150    150    150    230    230    230    230    230
 POST /api/1/discussions/5cf7d153dc47091406fe0665/                  10    100    110    110    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d154dc47097e573fa131/                  10    110    120    120    140    140    140    140    140    140
 POST /api/1/discussions/5cf7d156dc47091405a78005/                  10    100    100    110    110    110    110    110    110    110
 POST /api/1/discussions/5cf7d157dc47095ca2366744/                  10    140    160    160    170    170    170    170    170    170
 POST /api/1/discussions/5cf7d159dc47091406fe0666/                  10    120    120    120    120    130    130    130    130    130
 POST /api/1/discussions/5cf7d15adc47097e573fa132/                  10    180    280    310    320    620    620    620    620    620
 POST /api/1/discussions/5cf7d15cdc47091405a78006/                  10    190    190    320    320   1100   1100   1100   1100   1100
 POST /api/1/discussions/5cf7d160dc470914041794ec/                  10    100    100    110    110    750    750    750    750    750
 POST /api/1/discussions/5cf7d161dc47095ca2366745/                  10    100    100    110    110    130    130    130    130    130
 POST /api/1/discussions/5cf7d163dc47091406fe0667/                  10    110    110    110    120    320    320    320    320    320
 POST /api/1/discussions/5cf7d164dc47097e573fa133/                  10    100    110    110    110    110    110    110    110    110
 POST /api/1/discussions/5cf7d165dc47091405a78007/                  10    120    120    120    140    490    490    490    490    490
 POST /api/1/discussions/5cf7d167dc470914041794ed/                  10    100    100    110    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d168dc47091406fe0668/                  10    110    120    130    140    140    140    140    140    140
 POST /api/1/discussions/5cf7d169dc47091405a78008/                  10    140    150    200    210    270    270    270    270    270
 POST /api/1/discussions/5cf7d16bdc470914041794ee/                  10    110    110    110    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d16cdc47095ca2366747/                  10    100    110    110    120    120    120    120    120    120
 POST /api/1/discussions/5cf7d16edc47091406fe0669/                  10     99    100    100    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d16fdc47097e573fa134/                  10    120    120    130    150    530    530    530    530    530
 POST /api/1/discussions/5cf7d170dc470914041794ef/                  10    120    120    160    160    160    160    160    160    160
 POST /api/1/discussions/5cf7d172dc47095ca2366748/                  10    110    110    110    120    120    120    120    120    120
 POST /api/1/discussions/5cf7d173dc47091406fe066a/                  10    100    110    110    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d174dc47091405a78009/                  10    130    130    140    150    710    710    710    710    710
 POST /api/1/discussions/5cf7d176dc470914041794f0/                   4    180    180    270    270    270    270    270    270    270
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                             450    110    110    120    130    160    230    490    620   1100
```

## Rate Limit on, 500/minute (so limit not hit but counter is on)

```
Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 POST /api/1/datasets/                                               4    200    200    460    460    460    460    460    460    460
 POST /api/1/discussions/                                           38    130    140    150    200    220    280   1200   1200   1200
 POST /api/1/discussions/5cf7d32adc470907cc2f2132/                  10    200    200    240    270    300    300    300    300    300
 POST /api/1/discussions/5cf7d32cdc470907cdcc0b90/                  10    110    110    110    130    130    130    130    130    130
 POST /api/1/discussions/5cf7d32ddc470907cfedc2c2/                  10    120    120    130    130    270    270    270    270    270
 POST /api/1/discussions/5cf7d32edc470907ce74234a/                  10    170    190    290    350    490    490    490    490    490
 POST /api/1/discussions/5cf7d331dc470907cc2f2133/                  10    130    130    130    130    150    150    150    150    150
 POST /api/1/discussions/5cf7d332dc470907cdcc0b91/                  10    130    140    140    350   1400   1400   1400   1400   1400
 POST /api/1/discussions/5cf7d335dc470907cfedc2c3/                  10    130    130    160    170    200    200    200    200    200
 POST /api/1/discussions/5cf7d336dc470907d0ba25b5/                  10    120    140    140    160    300    300    300    300    300
 POST /api/1/discussions/5cf7d338dc470907ce74234b/                  10    130    130    130    150    160    160    160    160    160
 POST /api/1/discussions/5cf7d33adc470907cc2f2134/                  10    130    130    140    150    160    160    160    160    160
 POST /api/1/discussions/5cf7d33bdc470907cfedc2c4/                  10    200    260    280    290   1500   1500   1500   1500   1500
 POST /api/1/discussions/5cf7d33fdc470907ce74234c/                  10    120    130    130    130    140    140    140    140    140
 POST /api/1/discussions/5cf7d340dc470907cc2f2135/                  10    120    120    130    140    140    140    140    140    140
 POST /api/1/discussions/5cf7d341dc470907cdcc0b93/                  10    110    120    120    120    120    120    120    120    120
 POST /api/1/discussions/5cf7d343dc470907cfedc2c5/                  10    140    150    150    160    160    160    160    160    160
 POST /api/1/discussions/5cf7d344dc470907d0ba25b6/                  10    140    150    150    160    170    170    170    170    170
 POST /api/1/discussions/5cf7d346dc470907ce74234d/                  10    120    120    130    150   1000   1000   1000   1000   1000
 POST /api/1/discussions/5cf7d348dc470907cc2f2136/                  10    120    120    120    120    250    250    250    250    250
 POST /api/1/discussions/5cf7d34adc470907cdcc0b94/                  10    110    110    120    130    140    140    140    140    140
 POST /api/1/discussions/5cf7d34bdc470907cfedc2c6/                  10    120    120    130    130    130    130    130    130    130
 POST /api/1/discussions/5cf7d34cdc470907ce74234e/                  10    100    110    120    120    130    130    130    130    130
 POST /api/1/discussions/5cf7d34ddc470907cc2f2137/                  10    120    130    140    140    140    140    140    140    140
 POST /api/1/discussions/5cf7d34fdc470907cdcc0b95/                  10     97    100    110    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d350dc470907cfedc2c7/                  10    110    110    110    120    130    130    130    130    130
 POST /api/1/discussions/5cf7d351dc470907d0ba25b8/                  10    110    110    110    120    120    120    120    120    120
 POST /api/1/discussions/5cf7d352dc470907ce74234f/                  10    100    110    110    130    130    130    130    130    130
 POST /api/1/discussions/5cf7d354dc470907cc2f2138/                  10    110    110    110    110    110    110    110    110    110
 POST /api/1/discussions/5cf7d355dc470907cdcc0b96/                  10     99    100    100    100    120    120    120    120    120
 POST /api/1/discussions/5cf7d357dc470907cfedc2c8/                  10    110    110    110    110    130    130    130    130    130
 POST /api/1/discussions/5cf7d358dc470907d0ba25b9/                  10    100    100    100    110    130    130    130    130    130
 POST /api/1/discussions/5cf7d35adc470907cc2f2139/                  10    120    120    130    140    150    150    150    150    150
 POST /api/1/discussions/5cf7d35bdc470907cdcc0b97/                  10    110    120    120    140    150    150    150    150    150
 POST /api/1/discussions/5cf7d35cdc470907cfedc2c9/                  10    100    100    100    110    440    440    440    440    440
 POST /api/1/discussions/5cf7d35edc470907d0ba25ba/                  10    120    120    130    140    150    150    150    150    150
 POST /api/1/discussions/5cf7d35fdc470907ce742351/                  10    150    180    180    200    210    210    210    210    210
 POST /api/1/discussions/5cf7d361dc470907cc2f213a/                  10    100    110    110    110    130    130    130    130    130
 POST /api/1/discussions/5cf7d362dc470907cdcc0b98/                  10    140    140    150    170    310    310    310    310    310
 POST /api/1/discussions/5cf7d364dc470907cfedc2ca/                   2    110    110    110    110    110    110    110    110    110
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                             414    110    130    130    140    170    250    350    490   1500 
```

## Rate Limit on, 5/minute (limit hit right away)

```
Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 POST /api/1/datasets/                                              15    300    440   1100   2100   2500   7900   7900   7900   7900
 POST /api/1/discussions/                                           20    200    260    310    310    410   8100   8100   8100   8100
 POST /api/1/discussions/5cf7d459dc47091153798c52/                  10    100    150    200    200    260    260    260    260    260
 POST /api/1/discussions/5cf7d462dc4709176084bb06/                  10     84     85     86     86    120    120    120    120    120
 POST /api/1/discussions/5cf7d463dc47091761bffcb5/                  10     89     96    100    110    120    120    120    120    120
 POST /api/1/discussions/5cf7d465dc4709176084bb07/                  10    120    120    140    540   6800   6800   6800   6800   6800
 POST /api/1/discussions/5cf7d46ddc4709178aa83303/                  10    120    120   1200   1900   5900   5900   5900   5900   5900
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                              85    140    230    270    310   1100   2500   7900   8100   8100
```
```
Error report
 # occurrences      Error
--------------------------------------------------------------------------------------------------------------------------------------------
 5                  POST /api/1/discussions/5cf7d459dc47091153798c52/: "HTTPError('429 Client Error: TOO MANY REQUESTS for url: https://next.data.gouv.fr/api/1/discussions/5cf7d459dc47091153798c52/',)"
 10                 POST /api/1/discussions/5cf7d462dc4709176084bb06/: "HTTPError('429 Client Error: TOO MANY REQUESTS for url: https://next.data.gouv.fr/api/1/discussions/5cf7d462dc4709176084bb06/',)"
 10                 POST /api/1/discussions/5cf7d463dc47091761bffcb5/: "HTTPError('429 Client Error: TOO MANY REQUESTS for url: https://next.data.gouv.fr/api/1/discussions/5cf7d463dc47091761bffcb5/',)"
 10                 POST /api/1/discussions/5cf7d465dc4709176084bb07/: "HTTPError('429 Client Error: TOO MANY REQUESTS for url: https://next.data.gouv.fr/api/1/discussions/5cf7d465dc4709176084bb07/',)"
 10                 POST /api/1/discussions/5cf7d46ddc4709178aa83303/: "HTTPError('429 Client Error: TOO MANY REQUESTS for url: https://next.data.gouv.fr/api/1/discussions/5cf7d46ddc4709178aa83303/',)"
 15                 POST /api/1/discussions/: "HTTPError('429 Client Error: TOO MANY REQUESTS for url: https://next.data.gouv.fr/api/1/discussions/',)"
--------------------------------------------------------------------------------------------------------------------------------------------
```

## tldr;

There seems to be a performance hit when rate limiting, which makes sense since we need to store the counter. We probably need to do some more testing, but the gap isn't huge (414 objects created in one minute vs 450).

**Comment by abulte** [comment]
@noirbizarre me souffle, avec raison je pense, qu'on pourrait faire ça plutôt côté Nginx https://www.nginx.com/blog/rate-limiting-nginx/. Ce serait moins inquiétant côté perf et ça permettrait de ne pas faire grossir udata.

**Comment by abulte** [comment]
Let's do some nginx stuff instead (poke @sdelafond)
